core/new-gui/src/app/common/util/context.ts

```diff
@@ -0,0 +1,25 @@
+
+export function ContextManager<Context>(defaultContext: Context) {
+    abstract class ContextManager {
+        private static contextStack: Context[] = [defaultContext];
+        public static getContext() {
+            return this.contextStack[this.contextStack.length-1];
+        }
+        public static enter(context: Context) {
+            this.contextStack.push(context);
+        }
+        public static exit() {
+            this.contextStack.pop();
+        }
+        public static withContext<T>(context: Context, callable: () => T): T{
+            try {
+                this.enter(context);
+                return callable();
+            } finally {
+                this.exit();
+            }
+        }
+    }
+
+    return ContextManager;
+}
```

Comment:
ContextManager is my attempt to bring python's context managers to typescript. It can be used to:
 - bind and release resources reliably
 - act as \*global state\*, but only controlled from one place in the code. Using withContext, you can execute a function with some context, and any resulting function call in the stack can be aware of that context.

This was originally introduced right before I figured out async mode was a toggle-able option
My idea was to set the paper to async permanently, and create a sync/async boolean context. 

for every function that changes the jointpaper, check if they're called in an async context, and if not, manually sync the jointpaper to emulate the behavior of sychronous mode.